### PR TITLE
IGNITE-21566 Fix failed to get the primary replica exception.

### PR DESCRIPTION
**Pull request by sanpwc**
https://issues.apache.org/jira/browse/IGNITE-21566

### CODE REVIEW

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java [hunk 1]
```diff
@@ -495,6 +495,8 @@ private CompletableFuture<Replica> startReplicaInternal(
             TopologyAwareRaftGroupService raftClient,
             PendingComparableValuesTracker<Long, Void> storageIndexTracker
     ) {
+        LOG.info("Replica is about to start [replicationGroupId={}].", replicaGrpId);
```
[the remark this answers is not in the document]
Comment (sanpwc) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:498]: I don't think so, replica start is rather rare activity and while our system isn't that stable I'd like to have an ability to see given locks as often as possible.

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java [hunk 2]
```diff
@@ -511,10 +513,12 @@ private CompletableFuture<Replica> startReplicaInternal(
         CompletableFuture<Replica> replicaFuture = replicas.compute(replicaGrpId, (k, existingReplicaFuture) -> {
             if (existingReplicaFuture == null || existingReplicaFuture.isDone()) {
                 assert existingReplicaFuture == null || isCompletedSuccessfully(existingReplicaFuture);
+                LOG.info("Replica is started [replicationGroupId={}].", replicaGrpId);
 
                 return completedFuture(newReplica);
             } else {
                 existingReplicaFuture.complete(newReplica);
+                LOG.info("Replica is started, existing waiter was completed [replicationGroupId={}].", replicaGrpId);
```
[the remark this answers is not in the document]
Comment (sanpwc) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:521]: Basically we write such kind of logs for us or CS. Any suggestions about renaming?

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java [hunk 3]
```diff
@@ -779,6 +783,18 @@ private void idleSafeTimeSync() {
      * @return True if the replica is started.
      */
     public boolean isReplicaStarted(ReplicationGroupId replicaGrpId) {
+        CompletableFuture<Replica> replicaFuture = replicas.get(replicaGrpId);
+        return replicaFuture != null && isCompletedSuccessfully(replicaFuture);
```
[the remark this answers is not in the document]
Comment (sanpwc) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:787]: You are right. Actually I though that I'm using `CompletableFutures#isCompletedSuccessfully`. Will fix.

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java [hunk 4]
```diff
@@ -779,6 +783,18 @@ private void idleSafeTimeSync() {
...
+    }
+
+    /**
+     * Check if replica was touched by an any actor.
+     *
+     * @param replicaGrpId Replication group id.
+     * @return True if the replica was touched.
+     */
+    @TestOnly
+    public boolean isReplicaTouched(ReplicationGroupId replicaGrpId) {
```
[the remark this answers is not in the document]
Comment (sanpwc) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:797]: Will rephrase.

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (alievmirza) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:522]: I would add then something like "existing replica waiters were completed", but it is up to you.

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java [hunk 5]
```diff
@@ -779,6 +780,18 @@ private void idleSafeTimeSync() {
      * @return True if the replica is started.
      */
     public boolean isReplicaStarted(ReplicationGroupId replicaGrpId) {
+        CompletableFuture<Replica> replicaFuture = replicas.get(replicaGrpId);
+        return replicaFuture != null && isCompletedSuccessfully(replicaFuture);
+    }
+
+    /**
+     * Check if replica was touched by an any actor. Touched here means either replica replica creation or replica waiter registration.
```
Comment (alievmirza) [inline modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:788]: `replica` is written twice

**Review by alievmirza — APPROVED** [review]

**Review by vldpyatkov — APPROVED** [review]

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (sanpwc) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java]: Rephrased.

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java (hunk 5 above)
Comment (sanpwc) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java]: Fixed.

**Pull request merged**